### PR TITLE
<random>: Improve binomial_distribution accuracy

### LINKED ISSUE #1530: <random>: Poisson approximation for binomial_distribution is not accurate
For `binomial_distribution(t,p)` with `t` >= 25 and `t*p` < 1, it is approximated by a Poisson distribution. This is not very accurate, especially for smaller values of `t`.

**Command-line test case**
```cpp
c:\temp>type test.cpp
#include <cmath>
#include <iostream>
#include <random>
#include <vector>

using namespace std;

int main() {
    constexpr int it_max  = 100'000;
    constexpr int n       = 25;
    constexpr double mean = 0.99;
    constexpr double p    = mean / n;
    constexpr double var  = n * p * (1.0 - p);

    mt19937 gen;
    binomial_distribution<> dist(n, p);

    vector<int> counts(n + 1);
    for (int i = 0; i < it_max; ++i) {
        ++counts[static_cast<size_t>(dist(gen))];
    }

    double sample_mean = 0.0;
    for (size_t i = 1; i < counts.size(); ++i) {
        sample_mean += static_cast<double>(i) * counts[i];
    }
    sample_mean /= it_max;

    double sample_var = 0.0;
    for (size_t i = 0; i < counts.size(); ++i) {
        sample_var += counts[i] * pow(i - sample_mean, 2);
    }
    sample_var /= it_max - 1;

    cout << "Expected variance: " << var << endl
         << "Observed variance: " << sample_var << endl;

    return 0;
}

c:\temp>cl /EHsc test.cpp
Microsoft (R) C/C++ Optimizing Compiler Version 19.28.29617 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

test.cpp
Microsoft (R) Incremental Linker Version 14.28.29617.0
Copyright (C) Microsoft Corporation.  All rights reserved.

/out:test.exe
test.obj

c:\temp>.\test.exe
Expected variance: 0.950796
Observed variance: 0.985299
```

**Expected behavior**
Observed variance should be within 0.01 (~95% CI) of expected.

### PR DESCRIPTION
**Pull request by MattStephanson**
Fixes #1530.
* For small mean, replaces Poisson approximation with exact waiting time method based on cumulative sum of exponential variates.
* Uses cheap but rigorous approximation of exit condition for fast return on first iteration.

### CODE REVIEW

**Comment by MattStephanson** [comment]
BTW, for reference, the underlying algorithm is from ["Non-Uniform Random Variate Generation"](http://luc.devroye.org/rnbookindex.html), Luc Devroye, section X.4, "Second waiting time method".

File: stl/inc/random [hunk 1]
```diff
@@ -2524,8 +2524,30 @@ private:
...
+            // Events are rare, use waiting time method.
+            const _Ty1 _Rand = _NRAND(_Eng, _Ty1);
+            {
+                // The exit condition is -log(1 - _Rand)/t > -log(1-p), which is equivalent to _Rand > 1 - (1-p)^t. If
+                // we have a cheap upper bound for 1-(1-p)^t, we can exit early without having to call log. We use two
+                // such bounds, one that is tight for mean ~0 and another for mean ~1. In the first case, Bernoulli's
+                // inequality gives -1+p*t >= -(1-p)^t, so 1 - (1-p)^t <= p*t = mean. For the other bound, 1-(1-p)^t =
+                // 1-(1-p)(1-mean/t)^(t-1) <= 1-(1-p)(1-1/t)^(t-1) <= 1-(1-p)/e.
+                const _Ty1 _Ub =
+                    std::min(_Par0._Mean, _Ty1{3.678794411714423216e-1} * _Par0._Px + _Ty1{6.32120558828557678e-1});
```
Comment (statementreply) [inline stl/inc/random:2536]:
```suggestion
                    (_STD min)(_Par0._Mean, _Ty1{3.678794411714423216e-1} * _Par0._Px + _Ty1{6.32120558828557678e-1});
```

File: stl/inc/random [hunk 2]
```diff
@@ -2524,8 +2524,30 @@ private:
...
+            _Ty1 _Sum = -_CSTD log(_Ty1{1.0} - _Rand) / _Par0._Tx;
+            if (_Sum > -_Par0._Logp1)
+                return 0;
+            _Res = _Ty{1};
+
+            do {
+                _Sum += -_CSTD log(_Ty1{1.0} - _NRAND(_Eng, _Ty1)) / (_Par0._Tx - _Res);
+                ++_Res;
+            } while (_Sum <= -_Par0._Logp1 && _Res != _Par0._Tx);
+            return (_Sum > -_Par0._Logp1) ? static_cast<_Ty>(_Res - _Ty{1}) : _Par0._Tx;
```
Comment (statementreply) [inline stl/inc/random:2550]: Could this control flow be reordered to avoid comparing the final `_Sum` with `-_Par0._Logp1` twice and overcounting `_Res`?

**Comment by MattStephanson** [comment]
> There is an inconsistency wrt `Rand` and `_Rand`

I think I'm going to restore the original version anyway, because `1.0 - _Par0._Mean` makes it impossible to return anything other than 0 when `_Par0._Mean < DBL_EPSILON`.

File: stl/inc/random [hunk 3]
```diff
@@ -2524,8 +2524,26 @@ private:
...
+            {
+                // The exit condition is log(1 - _Rand)/t < log(1-p), which is equivalent to _Rand > 1 - (1-p)^t. If
+                // we have a cheap upper bound for 1-(1-p)^t, we can exit early without having to call log. We use two
+                // such bounds, one that is tight for mean ~0 and another for mean ~1. In the first case, Bernoulli's
+                // inequality gives -1+p*t >= -(1-p)^t, so 1 - (1-p)^t <= p*t = mean. For the other bound, 1-(1-p)^t =
+                // 1-(1-p)(1-mean/t)^(t-1) <= 1-(1-p)(1-1/t)^(t-1) <= 1-(1-p)/e.
+                const _Ty1 _Ub =
+                    (_STD min)(_Par0._Mean, _Ty1{3.678794411714423216e-1} * _Par0._Px + _Ty1{6.32120558828557678e-1});
+                if (_Rand > _Ub)
+                    return 0;
```
Comment (StephanTLavavej) [inline stl/inc/random:2538]: We conventionally use braces for all control flow.

File: tests/std/tests/GH_001530_binomial_accuracy/test.cpp [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    constexpr double mean = 0.99;
+    constexpr double p    = mean / n;
+    constexpr double var  = n * p * (1.0 - p);
+
+    mt19937 gen;
+    binomial_distribution<> dist(n, p);
+
+    vector<int> counts(n + 1);
+    for (int i = 0; i < it_max; ++i) {
+        ++counts[static_cast<size_t>(dist(gen))];
```
Comment (StephanTLavavej) [inline tests/std/tests/GH_001530_binomial_accuracy/test.cpp:23]: Technically should include `<cstddef>`.

**Review by StephanTLavavej — COMMENTED** [review]
Looks good to me - the logic here is significantly beyond my experience so I really appreciate the PR. :heart_eyes_cat: I will push changes for 2 trivial things I noticed.

**Review by StephanTLavavej — APPROVED** [review]

**Review by statementreply — CHANGES_REQUESTED** [review]
The proposed change doesn't correctly handle the case that _n_ * (1-_p_) < 1.

When _p_ > 0.5, except for the small _n_ case, the original algorithm generates _B_(_n_, _p_) as _n_ - _B_(_n_, 1-_p_). See `binomial_distribution::param_type::_Init()`:

``` cpp
_Pp    = _Px < 0.5 ? _Px : (1.0 - _Px);
_Mean  = _Tx * _Pp;
```

and the `return` below:

```cpp
return _Par0._Px == _Par0._Pp ? _Res : static_cast<_Ty>(_Par0._Tx - _Res);
```

Test case:

```cpp
#include <cmath>
#include <iostream>
#include <random>
#include <vector>

using namespace std;

int main() {
    constexpr int it_max  = 10'000'000;
    constexpr int n       = 25;
    constexpr double mean = n - 0.99;
    constexpr double p    = mean / n;
    constexpr double var  = n * p * (1.0 - p);

    mt19937 gen;
    binomial_distribution<> dist(n, p);

    vector<int> counts(n + 1);
    for (int i = 0; i < it_max; ++i) {
        ++counts[static_cast<size_t>(dist(gen))];
    }

    double sample_mean = 0.0;
    for (size_t i = 1; i < counts.size(); ++i) {
        sample_mean += static_cast<double>(i) * counts[i];
    }
    sample_mean /= it_max;

    cout << "Expected mean: " << mean << endl
         << "Observed mean: " << sample_mean << endl;

    double sample_var = 0.0;
    for (size_t i = 0; i < counts.size(); ++i) {
        sample_var += counts[i] * pow(i - sample_mean, 2);
    }
    sample_var /= it_max - 1;

    cout << "Expected variance: " << var << endl
         << "Observed variance: " << sample_var << endl;

    return 0;
}
```

Pre-PR output:

```
Expected mean: 24.01
Observed mean: 24.0104
Expected variance: 0.950796
Observed variance: 0.988557
```

Post-PR output:

```
Expected mean: 24.01
Observed mean: 0.990254
Expected variance: 0.950796
Observed variance: 0.950525
```

(Sorry I missed this in my previous review)

**Comment by MattStephanson** [comment]
@StephanTLavavej I've pushed a bugfix based on @statementreply's feedback after your approval.

File: tests/std/tests/GH_001530_binomial_accuracy/test.cpp [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    sample_mean /= it_max;
+
+    double sample_var = 0.0;
+    for (size_t i = 0; i < counts.size(); ++i) {
+        sample_var += counts[i] * pow(i - sample_mean, 2);
+    }
+    sample_var /= it_max - 1;
+
+    assert(abs(sample_mean / mean - 1.0) < 0.01);
+    assert(abs(sample_var / var - 1.0) < 0.01);
```
Comment (statementreply) [inline tests/std/tests/GH_001530_binomial_accuracy/test.cpp:39]:
```suggestion
    assert(abs(sample_mean / mean - 1.0) < tol);
    assert(abs(sample_var / var - 1.0) < tol);
```

Comment (MattStephanson) [reply tests/std/tests/GH_001530_binomial_accuracy/test.cpp:39]: 🤦‍♂️

**Review by StephanTLavavej — APPROVED** [review]

**Review by CaseyCarter — APPROVED** [review]
FYI: I've pushed a merge with `main`, and updated the list of libcxx test skips to indicate that my guess that #1530 caused `std/numerics/rand/rand.dis/rand.dist.bern/rand.dist.bern.bin/eval.PR44847.pass.cpp` to fail was wrong.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks for noticing and fixing this accuracy bug! :dart: :dart: :dart:

**Pull request merged**